pkg/ccl/changefeedccl/helpers_test.go

```diff
@@ -80,53 +81,137 @@ func readNextMessages(f cdctest.TestFeed, numMessages int, stripTs bool) ([]stri
+		}
+	}
+	return actual, nil
+}
+
+func stripTsFromPayloads(payloads []cdctest.TestFeedMessage) ([]string, error) {
+	var actual []string
+	for _, m := range payloads {
+		var value []byte
+		var message map[string]interface{}
+		if err := gojson.Unmarshal(m.Value, &message); err != nil {
+			return nil, errors.Newf(`unmarshal: %s: %s`, m.Value, err)
+		}
+		delete(message, "updated")
+		value, err := reformatJSON(message)
+		if err != nil {
+			return nil, err
 		}
+		actual = append(actual, fmt.Sprintf(`%s: %s->%s`, m.Topic, m.Key, value))
 	}
 	return actual, nil
 }
 
-func assertPayloadsBase(t testing.TB, f cdctest.TestFeed, expected []string, stripTs bool) {
+func extractUpdatedFromValue(value []byte) (float64, error) {
+	var updatedRaw struct {
+		Updated string `json:"updated"`
+	}
+	if err := gojson.Unmarshal(value, &updatedRaw); err != nil {
+		return -1, errors.Newf(`unmarshal: %s: %s`, value, err)
+	}
+	updatedVal, err := strconv.ParseFloat(updatedRaw.Updated, 64)
+	if err != nil {
+		return -1, errors.Wrapf(err, "error parsing updated timestamp: %s", updatedRaw.Updated)
+	}
+	return updatedVal, nil
+
+}
+
+func checkPerKeyOrdering(payloads []cdctest.TestFeedMessage) (bool, error) {
+	// map key to list of timestamp, ensure each list is ordered
+	keysToTimestamps := make(map[string][]float64)
+	for _, msg := range payloads {
+		key := string(msg.Key)
+		updatedTimestamp, err := extractUpdatedFromValue(msg.Value)
+		if err != nil {
+			return false, err
+		}
+		if _, ok := keysToTimestamps[key]; !ok {
+			keysToTimestamps[key] = []float64{}
+		}
+		if len(keysToTimestamps[key]) > 0 {
+			if updatedTimestamp < keysToTimestamps[key][len(keysToTimestamps[key])-1] {
+				return false, nil
+			}
+		}
+		keysToTimestamps[key] = append(keysToTimestamps[key], updatedTimestamp)
+	}
+	return true, nil
+}
+
+func assertPayloadsBase(
+	t testing.TB, f cdctest.TestFeed, expected []string, stripTs bool, perKeyOrdered bool,
+) {
 	t.Helper()
-	require.NoError(t, assertPayloadsBaseErr(f, expected, stripTs))
+	require.NoError(t, assertPayloadsBaseErr(f, expected, stripTs, perKeyOrdered))
 }
 
-func assertPayloadsBaseErr(f cdctest.TestFeed, expected []string, stripTs bool) error {
-	actual, err := readNextMessages(f, len(expected), stripTs)
+func assertPayloadsBaseErr(
+	f cdctest.TestFeed, expected []string, stripTs bool, perKeyOrdered bool,
+) error {
+	actual, err := readNextMessages(f, len(expected))
 	if err != nil {
 		return err
 	}
+
+	var actualFormatted []string
+	for _, m := range actual {
+		actualFormatted = append(actualFormatted, fmt.Sprintf(`%s: %s->%s`, m.Topic, m.Key, m.Value))
+	}
+
+	if perKeyOrdered {
+		ordered, err := checkPerKeyOrdering(actual)
+		if err != nil {
+			return err
+		}
+		if !ordered {
+			return errors.Newf("payloads violate CDC per-key ordering guaranteees\n  %s",
+				strings.Join(actualFormatted, "\n  "))
+		}
+	}
+
+	// strip timestamps after checking per-key ordering since check uses timestamps
+	if stripTs {
```

Comment:
move `stripTs` logic here since we might need to check per-key ordering first (which needs to be done before stripping the timestamps)